subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -667,6 +671,71 @@ static void dhcpv4_timeout(struct k_work *work)
 	}
 }
 
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
+
+static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
+				      enum net_dhcpv4_msg_type *msg_type, int length)
+{
+	struct net_dhcpv4_option_callback *cb, *tmp;
+	struct net_pkt_cursor backup;
+	uint8_t len;
+	uint8_t type;
+
+	if (length < 3) {
+		NET_ERR("Vendor-specific option parsing, length too short");
+		net_pkt_skip(pkt, length);
+		return -EBADMSG;
+	}
```

Comment:
The more I look at this block, the more concerned I feel. Are we really able to recover from this if we receive a perfectly formatted message but where length is 2, like when a malicious actor sends a hand crafted message and sets the length 2 where the protocol expects some bigger value. Now we move the cursor two bytes forward and continue parsing the data from wrong position, or did I understand this totally wrong?

Reply:
the `length` here is the whole length of the option 43. Therefore if we move the cursor 2 bytes, we are going to the next option in the dhcpv4 packet, skipping the rest of the option 43. And there either the next dhcpv4 option (not a encapsulated option) should be or the end off the dhcpv4 packet with `0xff`

Reply:
If the packet is malformed, then the parser will be out of sync and it is no point continuing parsing the message. I really think we should just discard it instead of trying to recover.

Reply:
What we are trying to recover from here is the case where option 43 is misconfigured, which could e.g. happen if the DHCP server was set up for devices from a different manufacturer.
According to [RFC2132](https://datatracker.ietf.org/doc/html/rfc2132#section-8.4), "Clients which do not receive desired vendor-specific information SHOULD make an attempt to operate without it, although they may do so (and announce they are doing so) in a degraded mode.". So trying to ignore option 43 would be in line with this.
If the option packet itself is corrupted, as could be the case in an attack scenario, that is something we might want to look into regarding the DHCP client in general. However, for option 43, I added most of these special cases into the tests. If there is any case you can think of that isn't covered by the tests, let me know. (The one descibed above is covered)

Reply:
Thanks for the clarifications and commenting the packet format in the test. I took your branch and experimented it a bit and it is super easy to get the dhcpv4 parser out of sync, but fortunately I did not see any real issues with it. Unknown options are just skipped, although it is probably possible to craft a message that allows valid options to be matched.

I think the dhcp option parser needs some looks to investigate if it is working properly, I can do that.

Reply:
> What we are trying to recover from here is the case where option 43 is misconfigured, which could e.g. happen if the DHCP server was set up for devices from a different manufacturer.

Hmm, but all we do here is checking if the payload length is smaller than 3 bytes? How would that prevent from "misconfigured" Vendor option (I assume you mean here that the vendor did not use the encapsulated TLVs internally?). 
That's why I was opting for treating this vendor-information option as opaque at the DHCP client level, and let the application interpret it (probably in par with class-identifier option content).

Reply:
Well, the easiest way of getting it out of sync are incorrect lengths. However, for my tests, I focused on the encapsulated option parser from this PR. And there, my main objective was making sure that for any incorrectly encoded option 43 the main parser would resume with the next option. For this I had to make the assumption that the lengths at the top level are correct. If they aren't it's not really related to this PR

Reply:
I don't think we should do parsing differently in vendor options compared to normal dhcpv4 options. For example if there is length error for normal options we just drop the packet immediately. So I am proposing these additional changes here:
```
diff --git a/subsys/net/lib/dhcpv4/dhcpv4.c b/subsys/net/lib/dhcpv4/dhcpv4.c
index 4f9de7a3fbe..c75e90d76f6 100644
--- a/subsys/net/lib/dhcpv4/dhcpv4.c
+++ b/subsys/net/lib/dhcpv4/dhcpv4.c
@@ -683,7 +683,6 @@ static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
 
        if (length < 3) {
                NET_ERR("Vendor-specific option parsing, length too short");
-               net_pkt_skip(pkt, length);
                return -EBADMSG;
        }
 
@@ -706,7 +705,6 @@ static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
                length--;
                if (length < len) {
                        NET_ERR("Vendor-specific option parsing, length too long");
-                       net_pkt_skip(pkt, length);
                        return -EBADMSG;
                }
                net_pkt_cursor_backup(pkt, &backup);
@@ -882,9 +880,13 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 #if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
                case DHCPV4_OPTIONS_VENDOR_SPECIFIC: {
                        if (!sys_slist_is_empty(&option_vendor_callbacks)) {
+                               int ret;
+
                                NET_DBG("options_vendor_specific");
-                               if (dhcpv4_parse_option_vendor(pkt, iface, msg_type, length) ==
-                                   -ENOBUFS) {
+
+                               ret = dhcpv4_parse_option_vendor(pkt, iface, msg_type,
+                                                                length);
+                               if (ret == -ENOBUFS || ret == -EBADMSG) {
                                        return false;
                                }
                        } else {
```
According to the RFC, all the options should have a length field. I did not see mentions what should be done if length is wrong, but I would imagine that it is quite reasonable to drop the packet if there is some error that could compromise the integrity of the message.

Reply:
We only return `-EBADMSG` when the length inside of one vender specific option is wrong. I.e. we would otherwise go to far with the cursor (outside of the DHCP 43 option). But if we just have a non encapsulated vendor option like `0x2b, 0x04, 0x03, 0x0F, 0x10, 0x01`, which according to the RFC is still valid, and still have `CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC` enabled we would drop the message, if we approve your change, because `0x0F` will be interpreted as a length of the first encapsulated option and it is longer than the field of the option 43, which is `0x04` long. In my opinion it is then wrong to drop the complete (and according to the RFC still valid)  packet. In that scenario we should instant  just ignore the content of the option 43 and go to the next option in the packet and continue. We can't know if the content of option 43 is encapsulated or not.

Reply:
My computer is a bit weird today. Assume that all my comments in this thread belong one message before where they are shown. Sorry about that.

Reply:
RFC 2132 
>Clients which do not receive desired vendor-specific information
   SHOULD make an attempt to operate without it, although they may do so
   (and announce they are doing so) in a degraded mode.

According to this I would interpret this to say that we should still not drop the packet (which means not getting a ip lease) if we don't get the encapsulated vendor options, which we desire and instant get non encapsulated vendor option, which we don't desire. We should still get the ip lease. the same should happen, if the content of the dhcp option 43 is malformed (= not desired), as long as length of the option 43 is right .

Reply:
Well, to be honest - I'm not convinced. It looks to me that all the logic around returning `EBADMSG` is just trying to circumvent the fact, that Vendor information option does not have a consistent format **and should be left** to the vendor-specific code to interpret (so not trying do to so in the generic library), as the RFC says:
> The information is an opaque object of n octets, **presumably interpreted by vendor-specific code on the clients
   and servers**.

Parsing the content in a generic part of the library, **assuming** that the option may have sub-TLVs (or may not), is IMO just a source of confusion in this thread.

Reply:
> Parsing the content in a generic part of the library, **assuming** that the option may have sub-TLVs (or may not), is IMO just a source of confusion in this thread.

Indeed, this seems to be the case here. Could we leave the vendor option parsing to application callback?

Reply:
If the vendor (in the RFC sense) decides to make use of encapsulated vendor specific options, they activate the corresponding Kconfig option. Then, they can use the provided functions that decode their encapsulated options in kernel space.
Like other functions provided by Zephyr, it is up to the vendor to activate them. For example, Zephyr provides lots of drivers for sensors, which are also not used by everyone - they also have to be activated by the vendor.
If a vendor doesn't want to use encapsulated vendor options, nothing prevents them from doing so. Without the Kconfig set, none of the code from this PR is compiled in.
 
With this PR, we are providing an easy and memory efficient way of using encapsulated vendor options that can be easily integrated into products. A main advantage of this is, that it makes it less likely for vendors to abuse "top level" options for their purposes.
Encapsulated vendor specific options are already being used in lots of places, SIP phones and thin clients being great examples. If someone were to implement one using Zephyr, this would be a great foundation for the DHCP side of things.
Due to their use by various products, they are already supported by various DHCP servers (e.g. MS DHCP, ISC Kea, dnsmasq) and wireshark has decoders for them built in. In the case of wireshark, option 60 is used to differentiate them. However, I would attribute this to the fact that wireshark has to be able to dissect encapsulated options for lots of vendors. On the other hand, it would certainly be possible to add a check for an expected value in option 60 before running this decoder. (obviously, that would depend on sending option 60 to the server first)

Reply:
> If a vendor doesn't want to use encapsulated vendor options, nothing prevents them from doing so. Without the Kconfig set, none of the code from this PR is compiled in.

Sure, this is business as usual and nothing new here. Both Robert and I are concerned about the implementation as it seems that the actual vendor extension parsing would better be done in the code that actually knows about the exact format of the data.

So would it make sense that instead of this where we try to parse semi opaque data,
```
	while (!net_pkt_read_u8(pkt, &type)) {
		if (type == DHCPV4_OPTIONS_END) {
			NET_DBG("Vendor-specific options_end");
			return 0;
		}
		length--;

		if (length <= 0) {
			NET_ERR("Vendor-specific option parsing, malformed option");
			return -EBADMSG;
		}

		if (net_pkt_read_u8(pkt, &len)) {
			NET_ERR("Vendor-specific option parsing, bad length");
			return -ENOBUFS;
		}
		length--;
		if (length < len) {
			NET_ERR("Vendor-specific option parsing, length too long");
			return -EBADMSG;
		}
		net_pkt_cursor_backup(pkt, &backup);

		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_vendor_callbacks, cb, tmp, node) {
			if (cb->option == type) {
				NET_ASSERT(cb->handler, "No callback handler!");

				if (net_pkt_read(pkt, cb->data, MIN(cb->max_length, len))) {
					NET_DBG("option vendor callback, read err");
					return -ENOBUFS;
				}

				cb->handler(cb, len, *msg_type, iface);
				net_pkt_cursor_restore(pkt, &backup);
			}
		}
		net_pkt_skip(pkt, len);
		length = length - len;
		if (length <= 0) {
			NET_DBG("Vendor-specific options_end (no code 255)");
			return 0;
		}
	}
```
we leave the parsing of the vendor specific data to the callback. We could do something like this (simplified example), so giving the whole vendor option directly to the callback to process it?
That way we also do not need to care what the format of the data is.
```
               bytes_copied = net_buf_linearize(user_supplied_buffer, user_supplied_buffer_len,
					pkt->frags, data_offset, length);

		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_vendor_callbacks, cb, tmp, node) {
			if (cb->option == type) {
				NET_ASSERT(cb->handler, "No callback handler!");

				cb->handler(cb, user_supplied_buffer, bytes_copied, iface);
			}
		}
```
WDYT?

Reply:
If I understand your proposal right, you just want to give back one callback for the whole content of the dhcp option 43. If that's what you mean, I don't see the benefit then instant just using the regular option call and register a callback for option 43. 
But that would still be possible to do with my commit. If the vendor doesn't want to use the encapsulated vendor specific option like my commit provides, he can still use the regular option callbacks and register a callback for the option 43 and handle the parsing in application code. But with my commit we enable the vendor to use the vendor specific options as they are recommended in the RFC, but hey don't have to.

Reply:
> But that would still be possible to do with my commit. If the vendor doesn't want to use the encapsulated vendor specific option like my commit provides, he can still use the regular option callbacks and register a callback for the option 43 and handle the parsing in application code. But with my commit we enable the vendor to use the vendor specific options as they are recommended in the RFC, but hey don't have to.

Indeed that is possible, so my proposal is not really needed. 

But imho my original concern is still valid, if there is a protocol error we should abort the parsing instead of trying to recover.

Reply:
> malformed encapsulated vendor specific option is still a valid vendor specific option, as long as the length of the whole option 43 is right

But what if this "malformed" option is indeed a valid vendor option from a different vendor that does not use encapsulated option encoding, but the first byte of their opaque data matches the suboption type we've registered a callback for? From the parser side, all could seem right, but we actually report rubbish to the application. Or even if some different vendor still uses encapsulated encoding, they could redefine the option codes as they wish, as it's allowed by RFC 1533:
> Codes other than 0 or 255 MAY be redefined by the vendor within the encapsulated vendor-specific extensions field.

That's why (I'm repeating myself), in my opinion any attempt to parse Vendor Specific option content (including assuming it uses encapsulated option encoding) should only take place after identifying whether we deal with a particualr vendor, as RFC 1533 says:
> The definition of this information is vendor specific. The vendor is indicated in the class-identifier option.

Otherwise, after enabling this feature, we're practically binding the generic implementation here to a single vendor. I may still work with other vendors, but could produce unexpected outcomes as described in the first paragraph.

Reply:
On more thought: assuming the DHCP server also sends back the clas identifier (don't know off the top of my head), what if there were only a single option 43 callback and rather than registering individual callbacks, there would just be buffers for the output of each suboption and a single callback, that callback could be only fired once an expected class identifier and a decodable option 43 have been received. Would that approach suit you better?